### DISCUSSION TITLE
Migrate @validator to @field_validator

### DISCUSSION THREAD

**Discussion by SergioEanX** [pydantic/pydantic#7693] [Pydantic V2]
I have to migrate `@validator` to `@field_validator` for the following code.

```
from pydantic import BaseModel, Field, ValidationInfo, field_validator
from pydantic_core import PydanticUndefined
from pydantic import ConfigDict
from uuid import uuid4
from typing import Union, Optional, Any


class newProject(BaseModel):
    identifier: Optional[str] = None
    proj_id: str = None
 
    v: Optional[str] = Field(default="1_0", alias="_v")


    @field_validator("proj_id", mode="plain", check_fields=True)
    @classmethod
    def set_id(cls, v: int, info: ValidationInfo) -> str:
        return v or "proj_number_-" + str(uuid4())[:6]

    model_config = ConfigDict(validate_assignment=True)


p = newProject(_v="1_1")
print(p.v)
print(p.proj_id)
```
 I obtain:

1_1
None

While I would like to have a random proj_id (as defined in set_id) if no field proj_id is not set while instantiating a newProject.
How to archive this?

**Comment by CharlesPerrotMinot** [comment]
I'm a bit confused by your code, because your field validator is set on `iemap_id`, but there's no such field in your Model. Did you mean `proj_id`? So `@field_validator("proj_id", mode="plain", check_fields=True)`?

From what I can tell, `@ field_validator` isn't called if the attribute (here  `proj_id`) isn't set in the model.
What I have done to migrate from a `@validator` (pydantic v1) in that case is to use `@model_validator`.

For example:
```
@model_validator(mode="before")
@classmethod
def set_id(cls, values: dict[str, Any]) -> dict[str, Any]:
    if not values.get("proj_id"):
        values["proj_id"] = "proj_number_-" + str(uuid4())[:6]
    return values
```

**Reply by SergioEanX** [reply]
Thanks, it worked. Sorry for the misunderstanding in the field name (it was a typo, now fixed).

**Comment by PHvL** [comment]
To set default values by a `field_validator` you could also annotate with `Field(validate_default=True)` to ensure the validator is called on the default `None` or instead of the validator set a default_factory: https://docs.pydantic.dev/latest/concepts/fields/#default-values

**Reply by CharlesPerrotMinot** [reply]
Nice one, I wasn't aware of this.
